### PR TITLE
Factor out cluster registry owned resource filtering into common function

### PR DESCRIPTION
**Pull request by stoader**
| Q               | A      |
| --------------- | ------ |
| Bug fix?        | no |
| New feature?    | no |
| API breaks?     | no |
| Deprecations?   | no |
| License         | Apache 2.0 |

### What's in this PR?

Factor out the logic of filtering the cluster registry owned resources from each controller into a common place.

### Why?

Simplify and reduce code size.

### Checklist

- [x] Implementation tested
- [x] Error handling code meets the [guideline](https://github.com/banzaicloud/developer-guide/blob/master/docs/coding-style/error-handling-guide.md)
- [x] Logging code meets the guideline

### CODE REVIEW

File: controllers/cruisecontroltask_controller.go [hunk 1]
```diff
@@ -22,16 +22,13 @@ import (
...
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/predicate"
 
-	"github.com/banzaicloud/koperator/pkg/util"
-
 	kafkav1beta1 "github.com/banzaicloud/koperator/api/v1beta1"
```
Comment (amberimam) [inline controllers/cruisecontroltask_controller.go:32]: This is not part of your change, but I was wondering if we can rename this to koperatorv1beta1?

File: controllers/kafkauser_controller.go [hunk 2]
```diff
@@ -128,7 +97,13 @@ func certificateSigningRequestFilter(log logr.Logger) predicate.Funcs {
 	}
 }
 
-func certificateSigningRequestMapper(obj client.Object) []ctrl.Request {
+type csrMapper struct {
+	client client.Reader
+	log    logr.Logger
+}
+
+// mapToKafkaUser maps CertificateSigningRequest event to KafkaUser reconcile event
```
Comment (amberimam) [inline controllers/kafkauser_controller.go:105]:
```suggestion
// mapToKafkaUser maps CertificateSigningRequest events to KafkaUser reconcile events
```

File: controllers/controller_common.go [hunk 3]
```diff
@@ -118,3 +122,23 @@ func applyClusterRefLabel(cluster *v1beta1.KafkaCluster, labels map[string]strin
 func SetNewKafkaFromCluster(f func(k8sclient client.Client, cluster *v1beta1.KafkaCluster) (kafkaclient.KafkaClient, func(), error)) {
 	newKafkaFromCluster = f
 }
+
+// SkipClusterRegistryOwnedResourcePredicate returns controller event filter that filters
+// out events triggered by Cluster Registry owned resources
+type SkipClusterRegistryOwnedResourcePredicate struct{}
+
+func (SkipClusterRegistryOwnedResourcePredicate) Create(e event.CreateEvent) bool {
+	return !util.ObjectManagedByClusterRegistry(e.Object)
```
Comment (amberimam) [inline controllers/controller_common.go:131]: Do we still need to do `object, err := meta.Accessor(e.Object)` before checking if the object is managed by cluster registry?

File: controllers/cruisecontroltask_controller.go (hunk 1 above)
Comment (stoader) [reply controllers/cruisecontroltask_controller.go:32]: in a separate PR

File: controllers/controller_common.go [hunk 4]
```diff
@@ -118,3 +122,23 @@ func applyClusterRefLabel(cluster *v1beta1.KafkaCluster, labels map[string]strin
 func SetNewKafkaFromCluster(f func(k8sclient client.Client, cluster *v1beta1.KafkaCluster) (kafkaclient.KafkaClient, func(), error)) {
 	newKafkaFromCluster = f
 }
+
+// SkipClusterRegistryOwnedResourcePredicate returns controller event filter that filters
+// out events triggered by Cluster Registry owned resources
+type SkipClusterRegistryOwnedResourcePredicate struct{}
+
+func (SkipClusterRegistryOwnedResourcePredicate) Create(e event.CreateEvent) bool {
```
[the remark this answers is not in the document]
Comment (stoader) [reply controllers/controller_common.go:130]: `WithEventFilter(SkipClusterRegistryOwnedResourcePredicate{})` see how it is used when setting up the controllers.

**Review by amberimam — APPROVED** [review]

File: controllers/kafkauser_controller.go [hunk 5]
```diff
@@ -138,10 +113,30 @@ func certificateSigningRequestMapper(obj client.Object) []ctrl.Request {
...
+	err := m.client.Get(ctx, client.ObjectKey{Namespace: namespace, Name: name}, &kafkaUser)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return []ctrl.Request{}
+		}
+		m.log.Error(err, "couldn't retrieve KafkaUser", "namespace", namespace, "name", name)
+		return []ctrl.Request{}
+	}
+
+	// skip reconciling KafkaUser if owned by Schema Registry
```
Comment (panyuenlau) [inline controllers/kafkauser_controller.go:131]:
```suggestion
	// skip reconciling KafkaUser if owned by Cluster Registry
```
This seems to be a typo

**Review by panyuenlau — APPROVED** [review]
LGTM

**Review by panyuenlau — APPROVED** [review]

**Review by amberimam — APPROVED** [review]

**Pull request merged**